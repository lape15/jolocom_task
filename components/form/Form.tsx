import { useState, useRef, useEffect } from "react";
import {
  StyleSheet,
  Text,
  View,
  TouchableOpacity,
  Animated,
} from "react-native";
import InputBox from "../inputs/Input";

type FormProps = {
  fields: {
    firstName: string;
    lastName: string;
    phoneNumber: string;
    email: string;
    age: string;
    isValid: boolean;
  };
  updateFields: Function;
  handleProfileSubmit: Function;
  showScreen: string;
};
const Form = (props: FormProps) => {
  const { fields, updateFields, handleProfileSubmit, showScreen } = props;
  const firstName = useRef<null | { focus: Function }>(null);
  const lastName = useRef<null | { focus: Function }>(null);
  const phoneNumber = useRef<null | { focus: Function }>(null);
  const email = useRef<null | { focus: Function }>(null);
  const age = useRef<null | { focus: Function }>(null);

  const fadeAnim = useRef(new Animated.Value(0)).current;
  const fadeIn = () => {
    // Will change fadeAnim value to 1 in 5 seconds
    Animated.timing(fadeAnim, {
      toValue: 1,
      duration: 800000,
      useNativeDriver: true,
      delay: 1000,
    }).start();
  };

  const fadeOut = () => {
    // Will change fadeAnim value to 0 in 3 seconds
    Animated.timing(fadeAnim, {
      toValue: 0,
      duration: 80000,
      useNativeDriver: true,
      delay: 1000,
    }).start();
  };
  useEffect(() => {
    if (showScreen === " ") {
      fadeIn();
    }
    return () => {
      fadeOut();
    };
  }, [showScreen]);

  const moveToNextField = (name: string) => {
    if (name === "lasttName") return lastName.current?.focus();
    if (name === "phoneNumber") return phoneNumber.current?.focus();
    if (name === "email") return email.current?.focus();
    if (name === "age") return age.current?.focus();
    if (name === "age") return firstName.current?.focus();
  };

  return (
    <Animated.View
      style={[
        styles.view,
        {
          transform: [
            { translateY: fadeAnim },
            // { scale: fadeAnim }
          ],
        },
      ]}
    >
      <InputBox
        value={fields.firstName}
        setValue={updateFields}
        name="firstName"
        placeholder="First name"
        keyboardType="default"
        refName="lastName"
        ref={firstName}
        moveToNextField={moveToNextField}
      />

      <InputBox
        value={fields.lastName}
        setValue={updateFields}
        name="lastName"
        placeholder="Last name"
        keyboardType="default"
        ref={lastName}
        refName="phoneNumber"
        moveToNextField={moveToNextField}
      />
      <InputBox
        value={fields.phoneNumber}
        setValue={updateFields}
        name="phoneNumber"
        keyboardType="numeric"
        placeholder="Phone number"
        ref={phoneNumber}
        refName="email"
        moveToNextField={moveToNextField}
      />
      <InputBox
        value={fields.email}
        setValue={updateFields}
        name="email"
        keyboardType="email-address"
        placeholder="Email"
        ref={email}
        refName="age"
        moveToNextField={moveToNextField}
      />
      <InputBox
        value={fields.age}
        setValue={updateFields}
        name="age"
        keyboardType="number-pad"
        placeholder="Age"
        ref={age}
        refName="firtName"
        moveToNextField={moveToNextField}
      />
      <View style={styles.btnView}>
        <TouchableOpacity
          onPress={() => handleProfileSubmit()}
          style={{ ...styles.btn, ...(fields.isValid && { opacity: 1 }) }}
          activeOpacity={1}
          disabled={!fields.isValid}
        >
          <Text style={styles.btnText}>Submit</Text>
        </TouchableOpacity>
      </View>
    </Animated.View>
  );
};

const styles = StyleSheet.create({
  view: {
    flex: 1,
    paddingTop: 20,
    marginHorizontal: 4,
    marginTop: 10,
    height: 750,
  },
  btnView: {
    flex: 1,
    marginTop: 20,
  },
  btn: {
    width: "100%",
    textAlign: "center",
    padding: 20,
    backgroundColor: "#E75480",
    color: "white",
    opacity: 0.3,
  },
  btnText: {
    textAlign: "center",
    color: "white",
  },
});

export default Form;
