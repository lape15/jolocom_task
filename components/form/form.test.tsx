/**
 * @jest-environment jsdom
 */
import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import Form from "./Form";

describe("form submits two answers", () => {
  const name = "firstName";
  const mockFn = jest.fn();
  const value = "firstName";
  jest.mock("react-native/Libraries/Animated/NativeAnimatedHelper");
  const fields = {
    firstName: "",
    lastName: "",
    phoneNumber: "",
    email: "",
    age: "",
    isValid: false,
  };

  it("matches snapshot", () => {
    const {
      getAllByA11yLabel,
      getByText,
      getByTestId,
      queryByTestId,
      queryAllByA11yLabel,
      toJSON,
    } = render(
      <Form
        fields={fields}
        handleProfileSubmit={mockFn}
        updateFields={mockFn}
        showScreen={"success"}
      />
    );

    const btn = getByTestId("submitInfo");
    fireEvent.press(btn);
    expect(queryByTestId("form")).toHaveLength(1);
    expect(toJSON()).toMatchSnapshot();
  });
});
