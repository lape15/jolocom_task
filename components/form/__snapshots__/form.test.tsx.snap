// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form submits two answers matches snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "height": 750,
      "marginHorizontal": 4,
      "marginTop": 25,
      "paddingTop": 30,
      "transform": Array [
        Object {
          "translateY": 0,
        },
      ],
    }
  }
  testID="form"
>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "margin": 10,
        "marginTop": 5,
        "padding": 10,
        "paddingTop": 30,
      }
    }
  >
    <TextInput
      accessibilityHint="input"
      accessibilityLabel="answer input"
      allowFontScaling={true}
      keyboardType="default"
      moveToNextField={[Function]}
      name="firstName"
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      placeholder="First name"
      placeholderTextColor="grey"
      refName="lastName"
      rejectResponderTermination={true}
      returnKeyType="done"
      setValue={[MockFunction]}
      style={
        Object {
          "backgroundColor": "rgba(126,94,129,0.25)",
          "borderColor": "rgba(126,94,129,0.1)",
          "borderRadius": 8,
          "borderStyle": "solid",
          "borderWidth": 1,
          "color": "white",
          "padding": 18,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </RCTSafeAreaView>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "margin": 10,
        "marginTop": 5,
        "padding": 10,
        "paddingTop": 30,
      }
    }
  >
    <TextInput
      accessibilityHint="input"
      accessibilityLabel="answer input"
      allowFontScaling={true}
      keyboardType="default"
      moveToNextField={[Function]}
      name="lastName"
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Last name"
      placeholderTextColor="grey"
      refName="phoneNumber"
      rejectResponderTermination={true}
      returnKeyType="done"
      setValue={[MockFunction]}
      style={
        Object {
          "backgroundColor": "rgba(126,94,129,0.25)",
          "borderColor": "rgba(126,94,129,0.1)",
          "borderRadius": 8,
          "borderStyle": "solid",
          "borderWidth": 1,
          "color": "white",
          "padding": 18,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </RCTSafeAreaView>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "margin": 10,
        "marginTop": 5,
        "padding": 10,
        "paddingTop": 30,
      }
    }
  >
    <TextInput
      accessibilityHint="input"
      accessibilityLabel="answer input"
      allowFontScaling={true}
      keyboardType="numeric"
      moveToNextField={[Function]}
      name="phoneNumber"
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Phone number"
      placeholderTextColor="grey"
      refName="email"
      rejectResponderTermination={true}
      returnKeyType="done"
      setValue={[MockFunction]}
      style={
        Object {
          "backgroundColor": "rgba(126,94,129,0.25)",
          "borderColor": "rgba(126,94,129,0.1)",
          "borderRadius": 8,
          "borderStyle": "solid",
          "borderWidth": 1,
          "color": "white",
          "padding": 18,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </RCTSafeAreaView>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "margin": 10,
        "marginTop": 5,
        "padding": 10,
        "paddingTop": 30,
      }
    }
  >
    <TextInput
      accessibilityHint="input"
      accessibilityLabel="answer input"
      allowFontScaling={true}
      keyboardType="email-address"
      moveToNextField={[Function]}
      name="email"
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Email"
      placeholderTextColor="grey"
      refName="age"
      rejectResponderTermination={true}
      returnKeyType="done"
      setValue={[MockFunction]}
      style={
        Object {
          "backgroundColor": "rgba(126,94,129,0.25)",
          "borderColor": "rgba(126,94,129,0.1)",
          "borderRadius": 8,
          "borderStyle": "solid",
          "borderWidth": 1,
          "color": "white",
          "padding": 18,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </RCTSafeAreaView>
  <RCTSafeAreaView
    emulateUnlessSupported={true}
    style={
      Object {
        "margin": 10,
        "marginTop": 5,
        "padding": 10,
        "paddingTop": 30,
      }
    }
  >
    <TextInput
      accessibilityHint="input"
      accessibilityLabel="answer input"
      allowFontScaling={true}
      keyboardType="number-pad"
      moveToNextField={[Function]}
      name="age"
      onBlur={[Function]}
      onChangeText={[Function]}
      onFocus={[Function]}
      onSubmitEditing={[Function]}
      placeholder="Age"
      placeholderTextColor="grey"
      refName="firtName"
      rejectResponderTermination={true}
      returnKeyType="done"
      setValue={[MockFunction]}
      style={
        Object {
          "backgroundColor": "rgba(126,94,129,0.25)",
          "borderColor": "rgba(126,94,129,0.1)",
          "borderRadius": 8,
          "borderStyle": "solid",
          "borderWidth": 1,
          "color": "white",
          "padding": 18,
        }
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </RCTSafeAreaView>
  <View
    style={
      Object {
        "flex": 1,
        "marginTop": 20,
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#E75480",
          "color": "white",
          "opacity": 0.3,
          "padding": 20,
          "textAlign": "center",
          "width": "100%",
        }
      }
      testID="submitInfo"
    >
      <Text
        style={
          Object {
            "color": "white",
            "textAlign": "center",
          }
        }
      >
        Submit
      </Text>
    </View>
  </View>
</View>
`;
